File: function/nodejs_14_x/src/handler/get_object_handler.ts
import { GetObjectContext, UserRequest } from '../s3objectlambda_event.types';
import { makeS3Request, applyRangeOrPartNumber } from '../request/utils';
import { AWSError } from 'aws-sdk/lib/error';
import ErrorCode from '../error/error_code';
import { Response, Headers } from 'node-fetch';
import { getErrorResponse, getResponseForS3Errors } from '../error/error_response';
import S3 from 'aws-sdk/clients/s3';
import { transformObject } from '../transform/s3objectlambda_transformer';
import { validate } from '../request/validator';
import { PromiseResult } from 'aws-sdk/lib/request';
import getChecksum from '../checksum/checksum';
import { headerToWgorParam, ParamsKeys } from '../response/param_transformer';

/**
 * Handles a GetObject request, by performing the following steps:
 * 1. Validates the incoming user request.
 * 2. Retrieves the original object from Amazon S3.
 * 3. Applies a transformation. You can apply your custom transformation logic here.
 * 4. Handles post-processing of the transformation, such as applying range or part numbers.
 * 5. Sends the final transformed object back to Amazon S3 Object Lambda.
 */

export default async function handleGetObjectRequest (s3Client: S3, requestContext: GetObjectContext, userRequest: UserRequest):
Promise<PromiseResult<{}, AWSError> | null> {
  // Validate user request and return error if invalid
  const errorMessage = validate(userRequest);
  if (errorMessage != null) {
    return getErrorResponse(s3Client, requestContext, ErrorCode.INVALID_REQUEST, errorMessage);
  }

  // Read the original object from Amazon S3
  const objectResponse = await makeS3Request(requestContext.inputS3Url, userRequest, 'GET');
  const originalObject = Buffer.from(await objectResponse.arrayBuffer());

  if (originalObject === null) {
    return getErrorResponse(s3Client, requestContext, ErrorCode.NO_SUCH_KEY, 'Requested key does not exist');
  }

  const responseHeaders = getResponseHeaders(objectResponse.headers);
  if (objectResponse.status >= 400) {
    // Errors in the Amazon S3 response should be forwarded to the caller without invoking transformObject.
    return getResponseForS3Errors(s3Client, requestContext, objectResponse, responseHeaders, originalObject);
  }

  if (objectResponse.status >= 300 && objectResponse.status < 400) {
    // Handle the redirect scenarios here such as Not Modified (304), Moved Permanently (301)
    return writeResponse(s3Client, requestContext, originalObject, responseHeaders, objectResponse);
  }

  // Transform the object
  const transformedWholeObject = transformObject(originalObject);

  // Handle range or partNumber if present in the request
  const transformedObject = applyRangeOrPartNumber(transformedWholeObject, userRequest);

  // Send the transformed object or error back to Amazon S3 Object Lambda
  if (transformedObject.hasError) {
    return getErrorResponse(s3Client, requestContext, ErrorCode.INVALID_REQUEST, String(transformedObject.errorMessage), responseHeaders);
  }
  if (transformedObject.object !== undefined) {
    return writeResponse(s3Client, requestContext, transformedObject.object, responseHeaders, objectResponse);
  }
  return null;
}

/**
 * Removes headers that will be invalidated by the transformation eg: Content-Length and ETag.
 */
function getResponseHeaders (headers: Headers): Headers {
  headers.delete('Content-Length');
  headers.delete('ETag');

  return headers;
}

/**
 * Send the transformed object back to Amazon S3 Object Lambda, by invoking the WriteGetObjectResponse API.
 */
async function writeResponse (s3Client: S3, requestContext: GetObjectContext, transformedObject: Buffer,
  headers: Headers, objectResponse: Response): Promise<PromiseResult<{}, AWSError>> {
  const { algorithm, digest } = getChecksum(transformedObject);

  const WGORParams = new Map<string, any>();

  // Create the Map with the params for WGOR
  headers.forEach((value, key) => {
    const paramKey = headerToWgorParam(key);
    if (ParamsKeys.includes(paramKey) && value !== '' && value !== null) {
      if (paramKey === 'LastModified') {
        WGORParams.set(paramKey, new Date(value));
      } else {
        WGORParams.set(paramKey, value);
      }
    }
  });

  console.log('Sending transformed results to the Object Lambda Access Point');
  return s3Client.writeGetObjectResponse({
    RequestRoute: requestContext.outputRoute,
    RequestToken: requestContext.outputToken,
    StatusCode: objectResponse.status,
    Body: transformedObject,
    Metadata: {
      'body-checksum-algorithm': algorithm,
      'body-checksum-digest': digest

    },
    ...Object.fromEntries(WGORParams)
  }).promise();
}

File: function/python_3_9/src/handler/get_object_handler.py
import checksum
import error
import requests
import transform
from request import utils, validator
from response import MapperResponse, range_mapper, part_number_mapper


def get_request_header(headers):
    """
    Get all headers that should be included in the pre-signed S3 URL. We do not add headers that will be
     applied after transformation, such as Range.
    :param headers: Headers from the GetObject request
    :return: Headers to be sent with pre-signed-url
    """
    new_headers = dict()
    headers_to_be_presigned = ['x-amz-expected-bucket-owner', 'x-amz-request-payer', 'If-Match',
                               'If-Modified-Since', 'If-None-Match', 'If-Unmodified-Since']
    for key, value in headers.items():
        if key in headers_to_be_presigned:
            new_headers[key] = value
    return new_headers


def get_object_handler(s3_client, request_context, user_request):
    """
    Handler for the GetObject Operation
    :param s3_client: s3 client
    :param request_context: GetObject request context
    :param user_request:  user request
    :return: WriteGetObjectResponse
    """
    # Validate user request and return error if invalid
    requests_validation = validator.validate_request(user_request)
    if not requests_validation.is_valid:
        return error.write_error_response(s3_client, request_context, requests.codes.bad_request,
                                        'InvalidRequest', requests_validation.error_msg)

    # Get the original object from Amazon S3
    s3_url = request_context["inputS3Url"]
    request_header = get_request_header(user_request["headers"])

    object_response = requests.get(s3_url, headers=request_header)

    # Check if the get original object request from S3 is successful
    if object_response.status_code != requests.codes.ok:
        # For 304 Not Modified, Error Message dont need to be send
        if object_response.status_code == requests.codes.not_modified:
            return s3_client.write_get_object_response(
                RequestRoute=request_context["outputRoute"],
                RequestToken=request_context["outputToken"],
                StatusCode=object_response.status_code,
            )
        return error.write_error_response_for_s3(s3_client,
                                                 request_context,
                                                 object_response)
    # Transform the object
    original_object = object_response.content

    transformed_whole_object = transform.transform_object(original_object)

    # Handle range or partNumber if present in the request
    partial_object_response = apply_range_or_part_number(transformed_whole_object, user_request)
    if partial_object_response.hasError:
        return error.write_error_response(s3_client, request_context, requests.codes.bad_request,
                                        'InvalidRequest', partial_object_response.error_msg)

    transformed_object = partial_object_response.object

    # Send the transformed object back to Amazon S3 Object Lambda
    transformed_object_checksum = checksum.get_checksum(transformed_object)
    return s3_client.write_get_object_response(RequestRoute=request_context["outputRoute"],
                                               RequestToken=request_context["outputToken"],
                                               Body=transformed_object,
                                               Metadata={
                                                   'body-checksum-algorithm': transformed_object_checksum.algorithm,
                                                   'body-checksum-digest': transformed_object_checksum.digest
                                               })


def apply_range_or_part_number(transformed_object, user_request):
    """
    Apply range or part number request to transformed object
    :param transformed_object: object that need to be apply range or part number
    :param user_request: request from the user
    :return: MapperResponse - response of the range or part number mapper
    """
    range_number = utils.get_range(user_request)
    part_number = utils.get_part_number(user_request)

    if part_number:
        return part_number_mapper.map_part_number(transformed_object, part_number)
    elif range_number:
        return range_mapper.map_range(transformed_object, range_number)
    return MapperResponse(hasError=False, object=transformed_object,
                          error_msg=None)
